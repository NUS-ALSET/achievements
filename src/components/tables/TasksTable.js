/**
 * @file TasksTable component module
 * @author Theodor Shaytanov <theodor.shaytanov@gmail.com>
 * @created 21.03.19
 */

import * as React from "react";
import PropTypes from "prop-types";

import Button from "@material-ui/core/Button";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";

import withStyles from "@material-ui/core/styles/withStyles";
import { Link } from "react-router-dom";

const styles = theme => ({
  link: {
    textDecoration: "none"
  },
  button: {
    margin: theme.spacing.unit
  }
});

class TasksTable extends React.PureComponent {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    tasks: PropTypes.object.isRequired,
    onDeleteClick: PropTypes.func.isRequired
  };

  render() {
    const { classes, onDeleteClick, tasks } = this.props;

    return (
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Task name</TableCell>
            <TableCell>Type</TableCell>
            <TableCell>Actions</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {!Object.keys(tasks).length ? (
            <TableRow>
              <TableCell colSpan={5}>Empty list</TableCell>
            </TableRow>
          ) : (
            Object.keys(tasks)
              .map(id => ({ ...tasks[id], id }))
              .map(task => (
                <TableRow key={task.id}>
                  <TableCell>{task.name}</TableCell>
                  <TableCell>{task.type}</TableCell>
                  <TableCell>
                    <Link className={classes.link} to={`/advanced/${task.id}`}>
                      <Button className={classes.button} variant="contained">
                        View
                      </Button>
                    </Link>
                    <Button
                      className={classes.button}
                      onClick={() => onDeleteClick(task.id)}
                      variant="contained"
                    >
                      Delete
                    </Button>
                  </TableCell>
                </TableRow>
              ))
          )}
        </TableBody>
      </Table>
    );
  }
}

export default withStyles(styles)(TasksTable);
