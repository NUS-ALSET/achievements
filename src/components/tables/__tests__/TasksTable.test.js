/* eslint-disable no-magic-numbers */
import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import sinon from "sinon";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";
import TableCell from "@material-ui/core/TableCell";
import TasksTable from "../TasksTable";
import { TableBody, TableRow } from "@material-ui/core";

describe("<TasksTable>", () => {
  let shallow;
  let mockDispatch;
  let tasks;

  beforeEach(() => {
    mockDispatch = sinon.spy();
    shallow = createShallow({
      dive: true
    });
    tasks = {
      task1: {},
      task2: {}
    };
  });

  it("renders correctly when there are multiple tasks", () => {
    const tree = renderer
      .create(
        <Router>
          <TasksTable onDeleteClick={mockDispatch} tasks={tasks} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render empty list", () => {
    const wrapper = shallow(
      <TasksTable onDeleteClick={mockDispatch} tasks={{}} />
    );

    expect(wrapper.find(TableBody).find(TableRow).length).toEqual(1);
  });

  it("should render tasks list", () => {
    const wrapper = shallow(
      <TasksTable onDeleteClick={mockDispatch} tasks={tasks} />
    );

    // Expect the wrapper object to be defined
    expect(wrapper.find(TableBody)).toBeDefined();
    expect(wrapper.find(TableBody).find(TableRow)).toHaveLength(
      Object.keys(tasks).length
    );
  });

  it("should render tasks without throwing", () => {
    const wrapper = shallow(
      <TasksTable onDeleteClick={mockDispatch} tasks={tasks} />
    );

    // Expect the wrapper object to be defined
    expect(wrapper.find(TableBody)).toBeDefined();
    expect(
      wrapper
        .find(TableBody)
        .find(TableRow)
        .find(TableCell)
    ).toHaveLength(6);
  });

  it("renders correctly when there are no tasks", () => {
    const tree = renderer
      .create(<TasksTable onDeleteClick={mockDispatch} tasks={{}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly when there are multiple tasks", () => {
    const tree = renderer
      .create(
        <Router>
          <TasksTable onDeleteClick={mockDispatch} tasks={tasks} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
