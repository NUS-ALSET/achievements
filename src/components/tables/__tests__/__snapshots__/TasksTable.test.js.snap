// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TasksTable> renders correctly when there are multiple tasks 1`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Task name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Type
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="TasksTable-link-1"
          href="/advanced/task1"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 TasksTable-button-2"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-24"
            >
              View
            </span>
            <span
              className="MuiTouchRipple-root-52"
            />
          </button>
        </a>
        <button
          className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 TasksTable-button-2"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-24"
          >
            Delete
          </span>
          <span
            className="MuiTouchRipple-root-52"
          />
        </button>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="TasksTable-link-1"
          href="/advanced/task2"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 TasksTable-button-2"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-24"
            >
              View
            </span>
            <span
              className="MuiTouchRipple-root-52"
            />
          </button>
        </a>
        <button
          className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 TasksTable-button-2"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-24"
          >
            Delete
          </span>
          <span
            className="MuiTouchRipple-root-52"
          />
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<TasksTable> renders correctly when there are multiple tasks 2`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Task name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Type
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="TasksTable-link-1"
          href="/advanced/task1"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 TasksTable-button-2"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-24"
            >
              View
            </span>
            <span
              className="MuiTouchRipple-root-52"
            />
          </button>
        </a>
        <button
          className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 TasksTable-button-2"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-24"
          >
            Delete
          </span>
          <span
            className="MuiTouchRipple-root-52"
          />
        </button>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="TasksTable-link-1"
          href="/advanced/task2"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 TasksTable-button-2"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-24"
            >
              View
            </span>
            <span
              className="MuiTouchRipple-root-52"
            />
          </button>
        </a>
        <button
          className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 TasksTable-button-2"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-24"
          >
            Delete
          </span>
          <span
            className="MuiTouchRipple-root-52"
          />
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<TasksTable> renders correctly when there are no tasks 1`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Task name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Type
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
        colSpan={5}
      >
        Empty list
      </td>
    </tr>
  </tbody>
</table>
`;
